Add EIP: Portable Smart Contract Accounts (#7405)
ethereum/EIPs: EIPS/eip-7405.md

---
eip: 7405
title: Portable Smart Contract Accounts
description: A solution working at the Proxy (ERC-1967) layer, allowing users to easily migrate their SCAs between different wallets.
author: Aaron Yee (@aaronyee-eth)
discussions-to: https://ethereum-magicians.org/t/erc-7405-portable-smart-contract-accounts/15236
status: Draft
type: Standards Track
category: ERC
created: 2023-07-26

> requires: 1967

Reviewer: You mention EIP-191 in the specification section, so it probably should be here too.

Author: Okay

---

## Abstract

> The proposed EIP introduces a standardized solution called "Portable Smart Contract Accounts" (PSCA) that aims to address the lack of portability and compatibility issues faced by Smart Contract Accounts (SCA) across different wallet providers. Based on **[ERC-1967](./eip-1967.md)**, the PSCA system allows users to easily migrate their SCAs between different wallets using new, randomly generated migration keys. This provides a similar experience to exporting an externally owned account (EOA) with a private key or mnemonic. The system ensures security by employing signatures and time locks, allowing users to verify and cancel migration operations during the lock period, thereby preventing potential malicious actions. PSCA offers a non-intrusive and cost-effective approach, enhancing the interoperability and composability within the Account Abstraction (AA) ecosystem.

Reviewer: Removing a few words:

```suggestion
Portable Smart Contract Accounts (PSCA) address the lack of portability and compatibility faced by Smart Contract Accounts (SCA) across different wallet providers. Based on [ERC-1967](./eip-1967.md), the PSCA system allows users to easily migrate their SCAs between different wallets using new, randomly generated migration keys. This provides a similar experience to exporting an externally owned account (EOA) with a private key or mnemonic. The system ensures security by employing signatures and time locks, allowing users to verify and cancel migration operations during the lock period, thereby preventing potential malicious actions. PSCA offers a non-intrusive and cost-effective approach, enhancing the interoperability and composability within the Account Abstraction (AA) ecosystem.
```

## Motivation

> With the introduction of the **[ERC-4337](./eip-4337.md)** standard, Account Abstraction (AA) related infrastructure and Smart Contract Accounts (SCA) have been widely adopted in the community. However, unlike externally owned accounts (EOA), SCAs have a more diverse code space, leading to varying contract implementations across different wallet providers. Consequently, the lack of portability for SCAs has become a significant issue, making it challenging for users to migrate their accounts between different wallet providers. While **[ERC-6900](./eip-6900.md)** proposed a modular approach for SCA accounts, it comes with higher implementation costs and specific prerequisites for wallet implementations.

Reviewer: ```suggestion
With the introduction of the [ERC-4337](./eip-4337.md) standard, AA related infrastructure and SCAs have been widely adopted in the community. However, unlike EOAs, SCAs have a more diverse code space, leading to varying contract implementations across different wallet providers. Consequently, the lack of portability for SCAs has become a significant issue, making it challenging for users to migrate their accounts between different wallet providers. While [ERC-6900](./eip-6900.md) proposed a modular approach for SCA accounts, it comes with higher implementation costs and specific prerequisites for wallet implementations.
```

> With the introduction of the **[ERC-4337](./eip-4337.md)** standard, Account Abstraction (AA) related infrastructure and Smart Contract Accounts (SCA) have been widely adopted in the community. However, unlike externally owned accounts (EOA), SCAs have a more diverse code space, leading to varying contract implementations across different wallet providers. Consequently, the lack of portability for SCAs has become a significant issue, making it challenging for users to migrate their accounts between different wallet providers. While **[ERC-6900](./eip-6900.md)** proposed a modular approach for SCA accounts, it comes with higher implementation costs and specific prerequisites for wallet implementations.

Reviewer: Small warning: if you link to another proposal, it limits how far your proposal can advance. For example, ERC-6900 is currently a Draft, which means this proposal cannot advance to Review until ERC-6900 does.

If you're okay waiting for ERC-6900, you can leave the reference in. Otherwise, just remove it.

Author: Okay

Considering that different wallet providers tend to prefer their own implementations or may expect their contract systems to be concise and robust, a modular system may not be universally applicable. The community currently lacks a more general SCA migration standard.

The goal of this proposal is to design a universal SCA migration mechanism that resembles the process of exporting an EOA account with a private key or mnemonic, as shown in the following diagram:

> ![Overview Diagram](../assets/eip-7405/overview-diagram.jpg)

Reviewer: If you can, please use an SVG. If you can't or prefer not to, a JPEG is fine too.

Author: Thanks. I'm just leaving a jpeg here.

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

### Terms

- Wallet Provider: A service provider that offers wallet services. SCA implementations among wallet providers are typically different, lacking compatibility with each other.
- Random Operator: A new, randomly generated migration mnemonic or private key used for each migration. The corresponding address of its public key is the random operator's address.

>     - If using a mnemonic, the derived migration private key follows the BIP-44 specification with the path **`m/44'/60'/0'/0/0'`**.

Reviewer: When linking to a BIP, please use the format specified in EIP-1:

```suggestion
    - If using a mnemonic, the derived migration private key follows the [BIP 44](https://github.com/bitcoin/bips/blob/55566a73f9ddf77b4512aca8e628650c913067bf/bip-0044.mediawiki) specification with the path **`m/44'/60'/0'/0/0'`**.
```

    - Sets the SCA account's lock status to false and deletes the record by calling **`IERC7405Registry.deleteMigrationData(randomOperator)`**.

### Storage Layout

To prevent conflicts in storage layout during migration across different wallet implementations, a Portable Smart Contract Account implementation contract:

- **MUST NOT** directly define state variables in the contract header.
- **MUST** encapsulate all state variables within a struct and store that struct in a specific slot. The slot index **SHOULD** be unique across different wallet implementations.

For slot index, we recommend calculating it based on the namespace and slot ID:

- The namespace **MUST** contain only [A-Za-z0-9_].
- Wallet providers' namespaces are **RECOMMENDED** to use snake_case, incorporating the wallet name and major version number, such as **`foo_wallet_v1`**.
- The slot ID for slot index **SHOULD** follow the format **`{namespace}.{customDomain}`**, for example, **`foo_wallet_v1.config`**.
- The calculation of the slot index is performed as **`bytes32(uint256(keccak256(slotID) - 1))`**.

## Rationale

> This proposal describes a solution working at the Proxy (ERC-1967) layer, providing a user experience similar to exporting an EOA account (using private keys or mnemonics). Considering that different wallet providers may have their own implementations, this solution imposes almost no requirements on the SCA implementation, making it more universally applicable and less intrusive with lower operational costs. Unlike a modular system operating at the "implementation" layer, both approaches can complement each other to further improve the interoperability and composability of the AA ecosystem.

Reviewer: This seems more like it belongs in the motivation section. The rationale should explain individual technical choices made within the EIP itself, while the motivation justifies the EIP as a whole.

Author: Okay, have rewritten the section as requested
